rechunker/types.py

```diff
@@ -1,31 +1,50 @@
 """Types definitions used by executors."""
-from typing import Any, Generic, Iterable, NamedTuple, Tuple, TypeVar
-
+from typing import Any, Generic, Iterable, NamedTuple, Optional, Tuple, TypeVar
 
+# TODO: replace with Protocols, once Python 3.8+ is required
+Array = Any
 ReadableArray = Any
 WriteableArray = Any
 
 
-class CopySpec(NamedTuple):
-    """Specifcation of how to copy between two arrays."""
+class ArrayProxy(NamedTuple):
+    """Representation of a chunked array for reads and writes.
 
-    source: ReadableArray
-    target: WriteableArray
-    chunks: Tuple[int, ...]
+    Attributes
+    ----------
+    array : array or None
+        If ``array`` is None, it represents an array without any on-disk
+        representation. Otherwise ``array`` should be an explicit array to read
+        and/or write to.
+    chunks : tuple
+        Chunks to use when reading/writing from this array. If ``array`` is
+        chunked, ``array.chunks`` will always even divide ``chunks``.
+    """
 
+    array: Optional[Array]
+    chunks: Tuple[int, ...]
 
-class StagedCopySpec:
-    """Specification of a copying process involving intermediate arrays.
 
-    The stages in a staged copy process must be completed in order. The
-    ``target`` of each stage can be assumed to correspond to the ``source`` of
-    the following stage.
+class CopySpec(NamedTuple):
+    """Specification for how to rechunk an array using a single intermediate.
+
+    Attributes
+    ----------
+    read : ArrayProxy
+        Read proxy with an ``array`` attribute that supports ``__getitem__``.
+    intermediate : ArrayProxy
+        Intermediate proxy with either an ``array`` that is either ``None``
```

Comment:
Thanks for the detailed explanation. I was confused about how direct copies would be represented.